src/generic/DirectSum.jl

```diff
@@ -20,6 +20,10 @@ base_ring(N::DirectSumModule{T}) where T <: RingElement = base_ring(N.m[1])
 
 base_ring(v::DirectSumModuleElem{T}) where T <: RingElement = base_ring(v.parent)
 
+is_free(M::DirectSumModule) = all(is_free, M.m)
```

Comment:
Actually, isn't this mathematically wrong? That is: if all summands are free, then M is free. But M might be free even though its summands are not (keyword: projective modules, etc.)

Reply:
@fieker In which situation was `is_free(::DirectSumModule)` missing?

Reply:
@thofma points out:
> If it is only applied to finitely generated/presented modules over fields or PIDs, then the method is correct: $M \oplus N$ is free if and only if $M$ and $N$ are free.

This is of course true, but is there a way we can use this knowledge to improve the code above? I mean we could certainly restrict it to modules over fields; or over AA "integers", and AA univariates polynomials over fields, and perhaps  a few more. But we can't do this generally here (e.g. not for fields over `Nemo.ZZRing`). But maybe we don't, it depends on what the application here is?

Reply:
No, but it is the same for all the functionality of things involving `FPModule`.

Reply:
Aha. The documentation indeed says:
```
# Free Modules and Vector Spaces

AbstractAlgebra allows the construction of free modules of any rank over any
Euclidean ring and the vector space of any dimension over a field. 
```
but that's the only place in `docs/src/free_module.md` mentioning Euclidean 
and it then proceeds to say
```
AbstractAlgebra provides generic types for free modules and vector spaces,
via the type `FreeModule{T}` for free modules, where `T`
is the type of the elements of the ring $R$ over which the module is built.
```
which to makes it not at all clear whether or not e.g. `FreeModule{T}` is only supposed to be used over Euclidean rings.

And `docs/src/direct_sum.md` does not mention Euclidean/principal/PID at all.


But `docs/src/module.md` then goes on to say this
```
The generic code provided by AbstractAlgebra will only work for modules over
euclidean domains.
```
which ought to settle it, except it *then* proceeds in the very next sentence to say
```
Free modules can be built over both commutative and noncommutative rings. Other
types of module are restricted to fields and euclidean rings.
```
which again leaves it completely unclear to me what is supposed to work in which case or not.

I think adding an `is_free` method that knowingly returns wrong results sometimes is problematic. At the very least its docstring then should contain a big warning?

Better would be if we could just restrict it -- e.g. if this only for direct sums over integers or fields it'd be fine. But that begs the question where is even used -- @fieker?


Alternatively I could also live with such a method:
```
function is_free(M::DirectSumModule)
  f = all(is_free, M.m)
  @req f || is_euclidean(base_ring(M)) "is_free for direct sums over non-euclidean domains not supported"
  return f
end

is_euclidean(::Field) = true
is_euclidean(::ZZRing) = true
is_euclidean(::PolyRing{<:Field}) = true
is_euclidean(::Ring) = error("cannot tell if this ring is Euclidean")
```